cmd/gardener-node-agent/app/app.go

```diff
@@ -113,23 +117,71 @@ func run(ctx context.Context, cancel context.CancelFunc, log logr.Logger, cfg *c
 		err        error
 	)
 
+	var mustBootstrap bool
 	if len(cfg.ClientConnection.Kubeconfig) > 0 {
 		restConfig, err = kubernetes.RESTConfigFromClientConnectionConfiguration(&cfg.ClientConnection, nil, kubernetes.AuthTokenFile)
 		if err != nil {
 			return fmt.Errorf("failed getting REST config from client connection configuration: %w", err)
 		}
 	} else {
-		var mustFetchAccessToken bool
-		restConfig, mustFetchAccessToken, err = getRESTConfig(log, cfg)
-		if err != nil {
-			return fmt.Errorf("failed getting REST config: %w", err)
+		var migrateKubeconfig bool
+		if features.DefaultFeatureGate.Enabled(features.NodeAgentAuthorizer) {
+			restConfig, mustBootstrap, err = getRESTConfig(log, fs, cfg)
+			if err != nil {
+				log.Info("Failed to get REST config for node-agent. Starting migration from access token kubeconfig")
+				migrateKubeconfig = true
```

Comment:
Unrelated errors could trigger the migration here. Shouldn't this be controlled by a dedicated return value?

Reply:
IO errors on the disk could trigger a migration. That's the only thing which could go wrong here (maybe except people deleting files from the disk). 
However, I'll check that with https://github.com/gardener/gardener/pull/10781#discussion_r1850717770 too.

Reply:
The trigger for starting migrations is the existence of kubeconfig/token file now.